books/kestrel/json-parser/parse-json.lisp

```diff
@@ -160,10 +163,70 @@
   (declare (xargs :guard (characterp char)))
   (member char *hex-digit-chars*))
 
+(defund hex-digit-to-nat (d)
+  (declare (xargs :guard (and (characterp d) (hex-digitp d))))
+  (let ((code (char-code d)))
+    (cond ((and (<= 48 code) (<= code 57))
+           (- code 48))
+          ((and (<= 65 code) (<= code 70))
+           (- code 55))
+          ((and (<= 97 code) (<= code 102))
+           (- code 87))
+          (t #x110000)))) ; an error value that should never happen
```

Comment:
Since this should be unreachable under the guard, you could consider returning `(prog2$ (impossible) 0)`, so you get a guard obligation from `(impossible)` and you still return an (irrelevant) natural number so this function unconditionally returns a natural number.